File: android/app/src/main/java/net/bradball/teetimecaddie/android/analytics/FirebaseErrorLogger.kt
package net.bradball.teetimecaddie.android.analytics

import com.google.firebase.crashlytics.FirebaseCrashlytics
import net.bradball.teetimecaddie.core.analytics.ErrorLogger

/**
 * Android [ErrorLogger] implementation backed by the native Firebase Crashlytics SDK.
 *
 * This is the platform-provided binding for error logging. It is created by the app and passed
 * into the SDK during initialization, keeping the shared code free of any Firebase dependency.
 */
class FirebaseErrorLogger : ErrorLogger {

    private val crashlytics: FirebaseCrashlytics
        get() = FirebaseCrashlytics.getInstance()

    override fun logException(throwable: Throwable) {
        crashlytics.recordException(throwable)
    }

    override fun logMessage(message: String) {
        crashlytics.log(message)
    }

    override fun recordStateValue(key: String, value: String) {
        crashlytics.setCustomKey(key, value)
    }

    override fun recordStateValue(key: String, value: Boolean) {
        crashlytics.setCustomKey(key, value)
    }

    override fun recordStateValue(key: String, value: Double) {
        crashlytics.setCustomKey(key, value)
    }

    override fun recordStateValue(key: String, value: Int) {
        crashlytics.setCustomKey(key, value)
    }

    override fun recordStateValue(key: String, value: Float) {
        crashlytics.setCustomKey(key, value)
    }

    override fun recordStateValue(key: String, value: Long) {
        crashlytics.setCustomKey(key, value)
    }

    override fun setUserId(userId: String) {
        crashlytics.setUserId(userId)
    }
}

File: ios/TeeTimeCaddie/TeeTimeCaddie/Source/analytics/FirebaseErrorLogger.swift
//
//  FirebaseErrorLogger.swift
//  TeeTimeCaddie
//
//  Platform (iOS) implementation of the shared ErrorLogger interface, backed by the
//  native Firebase Crashlytics SDK. Created by the app and passed into the SDK during
//  initialization, so the shared code carries no Firebase dependency.
//

import Foundation
import TeeTimeCaddieKit
import FirebaseCrashlytics

class FirebaseErrorLogger: ErrorLogger {

    private var crashlytics: Crashlytics { Crashlytics.crashlytics() }

    func logException(throwable: KotlinThrowable) {
        let error = NSError(domain: "TeeTimeCaddieKit", code: 0, userInfo: [
            NSLocalizedDescriptionKey: throwable.message ?? "Unknown Kotlin error",
            "kotlinExceptionType": String(describing: type(of: throwable))
        ])
        crashlytics.record(error: error)
    }

    func logMessage(message: String) {
        crashlytics.log(message)
    }

    // The shared ErrorLogger has six recordStateValue overloads (one per primitive type).
    // SKIE exposes them to Swift as clean type-based overloads (all labelled `value:`).
    func recordStateValue(key: String, value: Bool) {
        crashlytics.setCustomValue(value, forKey: key)
    }

    func recordStateValue(key: String, value: Double) {
        crashlytics.setCustomValue(value, forKey: key)
    }

    func recordStateValue(key: String, value: Float) {
        crashlytics.setCustomValue(value, forKey: key)
    }

    func recordStateValue(key: String, value: Int32) {
        crashlytics.setCustomValue(value, forKey: key)
    }

    func recordStateValue(key: String, value: Int64) {
        crashlytics.setCustomValue(value, forKey: key)
    }

    func recordStateValue(key: String, value: String) {
        crashlytics.setCustomValue(value, forKey: key)
    }

    func setUserId(userId: String) {
        crashlytics.setUserID(userId)
    }
}
